opencsp/common/lib/camera/ImageAcquisitionAbstract.py

```diff
@@ -1,8 +1,46 @@
 from abc import abstractmethod, ABC, abstractproperty
+import functools
+from typing import Callable, Optional
+
 import numpy as np
 
+import opencsp.common.lib.tool.exception_tools as et
+
 
 class ImageAcquisitionAbstract(ABC):
+    _instance: 'ImageAcquisitionAbstract' = None
+
+    @staticmethod
+    @functools.cache
+    def cam_options() -> dict[str, type['ImageAcquisitionAbstract']]:
+        """Defines camera objects to choose from (camera description, python type)"""
+        # import here to avoid circular reference
+        from opencsp.common.lib.camera.ImageAcquisition_DCAM_mono import ImageAcquisition as IA_DCAM_mono
+        from opencsp.common.lib.camera.ImageAcquisition_DCAM_color import ImageAcquisition as IA_DCAM_color
+        from opencsp.common.lib.camera.ImageAcquisition_MSMF import ImageAcquisition as IA_MSMF
+
+        cam_options: dict[str, type[ImageAcquisitionAbstract]] = {
+            'DCAM Mono': IA_DCAM_mono,
+            'DCAM Color': IA_DCAM_color,
+            'MSMF Mono': IA_MSMF,
+        }
+        return cam_options
+
+    def __init__(self):
+        if ImageAcquisitionAbstract._instance is None:
+            ImageAcquisitionAbstract._instance = self
+
+        self.on_close: list[Callable[[ImageAcquisitionAbstract], None]] = []
+
+    @classmethod
+    def instance(cls) -> Optional["ImageAcquisitionAbstract"]:
+        """Get the global ImageAcquisition instance, if one is available.
+
+        We use the singleton design pattern (single global instance) for this class because we don't expect there to be
+        more than one camera in the system.
+        """
+        return cls._instance

```

Comment:
So, just to make things more complicated, as of now, there can be multiple ImageAcquisition instances in a SofastSystemFringe, (sometimes we use redundant cameras in a Sofast setup). You are definitely more knowledgeable about singleton design architecture, so I expect you already accounted for this or there is an easy patch. Perhaps something like `self._instance` -> `self._instances` where `_instances` is a list of instances?? 

Note that in ImageAcquisition_DCAM_xxx.py classes, the constructor takes in an "instance" integer that selects which of the Basler cameras to connect to.
https://github.com/sandialabs/OpenCSP/blob/c3128f0789aac55a02de39d337a7a173744eb507/opencsp/common/lib/camera/ImageAcquisition_DCAM_color.py#L10

It's probably a good idea to keep this functionality... but probably not 100% necessary. We can discuss what you think the best idea is.

Reply:
Yeah, good call. Let's keep the multiple-camera functionality.

This doesn't really fit in the singleton design pattern any more. What we can do instead is use the multiton design pattern:

```python
class ImageAcquisitionAbstract:
    _instances: dict[int,ImageAcquisitionAbstract] = {}
    """ All instantiated camera instances. We use a dictionary to ensure that once an index is assigned,
    then all references to that same index will return the same camera (or None if the camera was closed). """
    _next_instance_idx: int = 0
    """ The index to next use when adding a camera to the _instances dict. """

    @staticmethod
    def get_instance(idx: int=0) -> ImageAcquisitionAbstract:
        if idx in ImageAcquisitionAbstract._instances:
            camera = ImageAcquisitionAbstract[idx]
            if camera.is_closed:
                return None
        return None

    @staticmethod
    def get_instances(subclass: type[ImageAcquisitionAbstract]=None) -> list[ImageAcquisitionAbstract]:
        ret: list[ImageAcquisitionAbstract] = []
        for idx in self._instances:
            camera = ImageAcquisitionAbstract.get_instance(idx)
            if camera is None:
                continue
            if subclass is not None and not isinstance(camera, subclass):
                continue
            ret.append(camera)
        return ret
```

Reply:
Sweet, nice work! Yes, multidon sounds like just what we want in this case.